### PR TITLE
Add parameter optimization.

### PR DESCRIPTION
**Pull request by akleeman**
Added the ability to create an arbitrary model, provide a dataset and an evaluation metric and be able to find the parameters which minimize the metric.

Includes:
- Addition of `tune.h` which deals with numerical optimization of model hyper parameters.
- A likelihood evaluation metric which can be used to do maximum likelihood optimization.
- Creates a new example `tune_example.cc`.

### CODE REVIEW

File: albatross/tune.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#include <map>
+#include <vector>
+#include "core/model.h"
+#include "nlopt.hpp"
+
+namespace albatross {
+
+std::vector<ParameterValue> transform_parameters(
+    const std::vector<ParameterValue>& x) {
+  std::vector<ParameterValue> transformed(x.size());
```
Comment (benjamin0) [inline albatross/tune.h]: nit: you could use `std::for_each` here and in the following functions below.

File: albatross/tune.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  TuneModelConfg(const RegressionModelCreator<Predictor> &model_creator_,
+                 const RegressionDataset<Predictor> &dataset_,
+                 const TuningMetric<Predictor> &metric_)
+      : model_creator(model_creator_), dataset(dataset_), metric(metric_){};
+};
+
+template <class Predictor>
+double objective_function(const std::vector<double> &x,
+                          std::vector<double> &grad, void *void_tune_config) {
```
Comment (benjamin0) [inline albatross/tune.h]: Is `grad` const? or is this the nlopt api?

Comment (akleeman) [reply albatross/tune.h]: That's the `nlopt` api.  If the optimization algorithm you choose requires gradients you're expected to have filled them in hence the non const.

File: albatross/tune.h (hunk 1 above)
Comment (akleeman) [reply albatross/tune.h]: done (though I had to use `transform` instead of `for_each` to avoid a copy).

**Pull request merged**